### PR TITLE
fix!: change persistent directory name to deterministic value

### LINKED ISSUE #1998: Remove ambiguity from the pvc names
### What's your CLI version?

0.85.53

### Description & steps to reproduce

When I create a new pvc with persistent volumes in kubernetes, I get random pvc names.

```sh
k -n kt-bitter-river get pvc
NAME                                                            STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
enclave-data                                                    Bound    pvc-c79fde18-dc50-4611-ae6b-b547e799d786   1Gi        RWO            hostpath       8m23s
service-persistent-directory-716dde3efcfe39cc9e6c7d4a596ca0eb   Bound    pvc-f9d54b5b-2e80-4e70-ac4f-b4e95480b21f   1Gi        RWO            hostpath       6m6s
service-persistent-directory-d06739f2d83303f42d83f3fd34477985   Bound    pvc-cae66c5f-e737-4dbd-b858-98113e88aec1   1Gi        RWO            hostpath       6m17s
service-persistent-directory-d92433dfcc7b10c1b69b5d8d752b35d9   Bound    pvc-521ea91f-9d50-478e-9620-d95a6743824d   1Gi        RWO            hostpath       5m48s
```

### Desired behavior

Ideally you would want to have the `persistent_key` as the name for the pvc. 

Currently you can only find this with pvc describe:

```sh
k -n kt-bitter-river describe pvc service-persistent-directory-716dde3efcfe39cc9e6c7d4a596ca0eb
Name:          service-persistent-directory-716dde3efcfe39cc9e6c7d4a596ca0eb
Namespace:     kt-bitter-river
StorageClass:  hostpath
Status:        Bound
Volume:        pvc-f9d54b5b-2e80-4e70-ac4f-b4e95480b21f
Labels:        kurtosistech.com/app-id=kurtosis
               kurtosistech.com/enclave-id=2fc69ee037e64f25991462b819906a44
               kurtosistech.com/guid=716dde3efcfe39cc9e6c7d4a596ca0eb
               kurtosistech.com/id=cl-2-prysm-geth
```

found under `kurtosistech.com/id=cl-2-prysm-geth`

### What is the severity of this bug?

Critical; I am blocked and Kurtosis is unusable for me because of this bug.

### What area of the product does this pertain to?

Frontend: the Graphical User Interface, including the Enclave Manager

### PR DESCRIPTION
**Pull request by h4ck3rk3y**
Closes #1998

### CODE REVIEW

**Comment by h4ck3rk3y** [comment]
Need to add a validator that asserts 1035!

**Comment by h4ck3rk3y** [comment]
@leoporoli this is WIP but this is technically a breaking change as earlier `persitent_data` would be a valid key but now as we label the pvc with the `persitent_key` it has to be 1035 comptaible(as we enforce this on all object names)

there is an argument to be made that the user shouldn't be using `kubectl get pvc --all-namespaces` and use KT instead; but this doesn't seem that harmful and this is what we already do for service names

File: container-engine-lib/lib/backend_interface/objects/service_directory/persistent_directories.go [hunk 1]
```diff
@@ -1,5 +1,25 @@
 package service_directory
 
+import "regexp"
+
+const (
+	// PersistentKeyRegex implements RFC-1035 for naming persistent directory keys, namely:
```
Comment (leoporoli) [inline container-engine-lib/lib/backend_interface/objects/service_directory/persistent_directories.go:6]: great comment!

File: container-engine-lib/lib/backend_interface/objects/service_directory/persistent_directories.go [hunk 2]
```diff
@@ -1,5 +1,25 @@
...
+	// * contain at most 63 characters
+	// * contain only lowercase alphanumeric characters or '-'
+	// * start with an alphabetic character
+	// * end with an alphanumeric character
+	// The adoption of RFC-1035 is to maintain compatability with current Kubernetes service and pod naming standards:
+	// We use this over RFC-1035 as persistent data require 1035 to be followed
+	// https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names
+	// https://kubernetes.io/docs/concepts/services-networking/service/
+	PersistentKeyRegex            = "[a-z]([-a-z0-9]{0,61}[a-z0-9])?"
+	WordWrappedPersistentKeyRegex = "^" + PersistentKeyRegex + "$"
```
Comment (leoporoli) [inline container-engine-lib/lib/backend_interface/objects/service_directory/persistent_directories.go:14]: I like how you used a wrapper here

File: container-engine-lib/lib/backend_interface/objects/service_directory/persistent_directories.go [hunk 3]
```diff
@@ -17,3 +37,7 @@ func NewPersistentDirectories(persistentDirectories map[string]PersistentDirecto
 		ServiceDirpathToPersistentDirectory: persistentDirectories,
 	}
 }
+
+func IsPersistentKeyValid(persistentKey DirectoryPersistentKey) bool {
```
Comment (leoporoli) [inline container-engine-lib/lib/backend_interface/objects/service_directory/persistent_directories.go:39]: I like that you put the validation inside the object, good way to centralize this logic.

File: core/server/api_container/server/startosis_engine/kurtosis_instruction/add_service/add_service_shared.go [hunk 4]
```diff
@@ -117,6 +126,16 @@ func invalidServiceNameErrorText(
 	)
 }
 
+func invalidPersistentKeyErrorText(
+	persistentKey service_directory.DirectoryPersistentKey,
+) string {
+	return fmt.Sprintf(
+		"Persistent Key '%v' is invalid as it contains disallowed characters. Persistent Key must adhere to the RFC 1035 standard, specifically implementing this regex and be 1-63 characters long: %s. This means the service name must only contain lowercase alphanumeric characters or '-', and must start with a lowercase alphabet and end with a lowercase alphanumeric character.",
```
Comment (leoporoli) [inline core/server/api_container/server/startosis_engine/kurtosis_instruction/add_service/add_service_shared.go:137]: nice

**Review by leoporoli — APPROVED** [review]

File: container-engine-lib/lib/backend_interface/objects/service_directory/persistent_directories.go (hunk 1 above)
Comment (h4ck3rk3y) [inline container-engine-lib/lib/backend_interface/objects/service_directory/persistent_directories.go:6]: Perhaps assert uniquenesss

File: core/server/api_container/server/startosis_engine/kurtosis_instruction/add_service/add_service_shared.go [hunk 5]
```diff
@@ -71,12 +71,15 @@ func validateSingleService(validatorEnvironment *startosis_validator.ValidatorEn
 		return startosis_errors.NewValidationError(invalidServiceNameErrorText(serviceName))
 	}
 
-	// TODO perhaps add AddPersistentKey DoesPersistentKeyExist methods to the validator environment checks for it  here
 	if persistentDirectories := serviceConfig.GetPersistentDirectories(); persistentDirectories != nil {
 		for _, directory := range persistentDirectories.ServiceDirpathToPersistentDirectory {
 			if !service_directory.IsPersistentKeyValid(directory.PersistentKey) {
 				return startosis_errors.NewValidationError(invalidPersistentKeyErrorText(directory.PersistentKey))
 			}
+			if validatorEnvironment.DoesPersistentKeyExist(directory.PersistentKey) == startosis_validator.ComponentCreatedOrUpdatedDuringPackageRun {
```
Comment (leoporoli) [inline core/server/api_container/server/startosis_engine/kurtosis_instruction/add_service/add_service_shared.go:79]: what happens if the value is `ComponentExistedBeforePackageRun`? shouldn't it also fail?

Comment (h4ck3rk3y) [reply core/server/api_container/server/startosis_engine/kurtosis_instruction/add_service/add_service_shared.go:79]: This cache won't ever get into that state, as we aren't persisting the value over every `NewValidatorEnvironment` creation, I have added that and I am tracking that as a TODO

**Pull request merged**